### ISSUE TITLE
_foreach_addc_

### ISSUE THREAD

**Issue by ad8e** [pytorch/pytorch#129329]
### 🚀 The feature, motivation and pitch

On the surface, I'd like to call `_foreach_addc_(list of tensors, list of tensors, 1D-tensor)`. This is because I am adding a gradient (argument 2) with per-layer learning rates (argument 3) to the params (argument 1). The per-layer learning rate is the average size of the weight in the layer. This technique is used by Adafactor, and has been seen before in Adam as "AdamWScale": https://github.com/PiotrNawrot/nanoT5/blob/1c82d67bf8dea635be68a3b2a68a43b68b665193/nanoT5/utils/copied_utils.py#L243

Reasoning more generally about APIs:

Per-layer LRs are single floats per parameter tensor. So are things like vector norms. I assume that `_foreach` ops are slower than regular tensor ops on GPU. So these collections of single floats should be 1D tensors, rather than Lists of tensors. For example, AdamW stacks lists of 0D tensors into a 1D tensor: https://github.com/pytorch/pytorch/blob/dadc0ed4c8ea0cd46402d7ff39415b3464b121cd/torch/optim/adamw.py#L589 I would like to use a 1D tensor as the scaling parameter in `_foreach_addc_`, just as AdamW does for `_foreach_addcmul_`. 

It also appears AdamW is hosting the tensor LR on CPU to reduce the cost of re-stacking this tensor from a list every step. I'm thinking I will instead assume the list of tensors never changes, and keep a fixed-size tensor of LRs.

Maybe I should want `_foreach_norm` to return a 1D tensor, instead of a tuple of tensors?

### Alternatives

I could do `_foreach_mul` then `_foreach_add`, with lists of tensors.

### Additional context

I'm investigating optimizer speed because I had a 20 TFLOPS decrease when going from 8xH100 FSDP to 1xH100, at 100M scale. That was unintuitive, because there's less comms. My guess is that increased optimizer work is the cause, although this issue is not related, since it addresses only fixed launch overhead.

**Comment by janeyx99** [comment]
> I assume that _foreach ops are slower than regular tensor ops on GPU.

I'm curious if you've benchmarked this as I'm not sure how big a difference a 1D Tensor vs a list of ScalarTensors would be. For `xs + a * ys`, where `xs` and `ys` are TensorLists, we already have the following schemas for:

`a` is a python Scalar or a ScalarTensor: _foreach_add.List(Tensor[] self, Tensor[] other, *, Scalar alpha=1) -> Tensor[]
`a` is a list of Tensors: _foreach_addcmul.Scalar(Tensor[] self, Tensor[] tensor1, Tensor[] tensor2, Scalar value=1) -> Tensor[]

It is true we do not have a schema where `a` is a 1-dimensional Tensor, but perf should be similar to using a list of ScalarTensors with `_foreach_addcmul`. If performance is the main ask, we would prefer making that overload as speedy as possible over adding new ops.

> I'm investigating optimizer speed

This is awesome, I'd love to improve any gaps you see. Looking holistically at this issue, I see you also mentioned `_foreach_norm` returning a 1D Tensor. The premise that 1D Tensors is faster than a List of ScalarTensors is interesting but not yet proven and I suspect we can get similar speeds through other means. One blocker I already see with changing the return type of _foreach_norm to a 1D Tensor is that it would break the semantics of "foreachifying" an op, which promises to just listify the inputs and outputs.

**Comment by ad8e** [comment]
Yeah, my main consideration is perf. Here's a small benchmark of `_foreach` vs 1D tensors, on an A40:
```
A = [torch.randn((1), device="cuda", dtype=torch.float16) for _ in range(N)]
B = [torch.randn((1), device="cuda", dtype=torch.float16) for _ in range(N)]
@torch.compile
def f():
    return torch._foreach_add(A, B)

N=1024: 0.34MF/s
N=2048: 0.30MF/s


A = torch.randn((N), device="cuda", dtype=torch.float16)
B = torch.randn((N), device="cuda", dtype=torch.float16)
@torch.compile
def f():
    return A + B

N=1024: 244.27MF/s
N=2048: 435.37MF/s
N=3072: 640.00MF/s
N=4096: 870.75MF/s
N=5120: 1059.60MF/s
N=6144: 1306.12MF/s
N=7168: 1513.51MF/s
```
So adding 1D tensors is 1000x faster than `_foreach_add`ing lists of tensors. Each element in the `_foreach` list adds 3 μs. My 10M transformer has ~36 tensors, so it's 0.1 ms per operation. A full step is 100 ms.

I don't know the considerations around these `_foreach` ops, so I can't say that turning all the scalar lists into 1D tensors would be problem-free. Here's part of what I'm doing:

```
# later, I may also use tensordot for per-head normalization in attention, and per-vocab in the embedding layer
dot_product = [torch.dot(normalized_exp_avg[i].view(-1), params[i].view(-1)) for i in range(len(active_p))]
current_norms = torch._foreach_norm(params)
torch._foreach_div_(dot_product, torch._foreach_pow(current_norms, 2.0))
torch._foreach_add_(dot_product, 1.0)
torch._foreach_mul_(initial_norm, dot_product)
ratio = torch._foreach_div(initial_norm, current_norms)
torch._foreach_mul_(params, ratio)
```
With 1D tensors replacing scalar lists, it would look like:
```
dot_product = torch._foreach_dot_returning_1D_tensor(normalized_exp_avg, params)
current_norms = torch._foreach_norm_returning_1D_tensor(params)
dot_product /= current_norms ** 2
dot_product += 1.0
initial_norm *= dot_product
torch._foreach_mul_(params, initial_norm / current_norms)
```

A 3-step trace.json of my model is here (not including the above code): https://drive.google.com/file/d/1vT85V1q9GPqiNbEPkZ9Djl0VEoh7qia-/view?usp=sharing It's a 10M transformer with 8604 tokens per batch. The optimizer takes a small amount of time. I notice the `_foreach_addcmul` is split up into multiple kernels, so it takes the same amount of time as the for-loop `[torch.atan2(...) for i in range(...)]` next to it. Speedups in MTA would be nice, but not impactful.

**Comment by janeyx99** [comment]
@ad8e The benchmarking example you show is expected, since the foreach op would have to go through dispatch overhead + foreach checks, and then a particular pattern for kernel launch that the vanilla add would avoid by simply going through TensorIterator. Hence, this isn't exactly a fair comparison for the requested addc op that would already need to go through foreach dispatch for the Tensor xs + ys. Maybe a more fair look will be looking at foreach_addcdiv and testing between the ScalarList and Tensor overloads?
![image](https://github.com/pytorch/pytorch/assets/31798555/8a3e6f5d-818b-4b01-b9be-0a86b85320dc)

I haven't gotten a chance to mathematicize your code sample yet, but to confirm the use case--this is outside of the optimizer step?

**Comment by ad8e** [comment]
> I haven't gotten a chance to mathematicize your code sample yet, but to confirm the use case--this is outside of the optimizer step?

The code sample I posted is inside my optimizer step. It does math to handle various boring and standard things in a better way. So having tensors replace lists would help not just in the _foreach ops, but also in the surrounding tensor math for the constants for these ops.

> Maybe a more fair look will be looking at foreach_addcdiv and testing between the ScalarList and Tensor overloads?

Right; my benchmark is not applicable to using a tensor as c, only for lists vs tensors in general.

Here's a benchmark for the tensor as c:

```
# everything is no-grad
A = [torch.rand((1,), device="cuda", dtype=torch.float16) for _ in range(N)]
B = [torch.rand((1,), device="cuda", dtype=torch.float16) for _ in range(N)]
C = [torch.ones((1,), device="cuda", dtype=torch.float16) for _ in range(N)]
# D = [1.0 for _ in range(N)] # turned on for list of scalars
# D = torch.rand(N, device="cpu", dtype=torch.float16) # turned on for tensor

# @torch.compile
def f():
	return torch._foreach_addcdiv(A, B, C, D)


GPU name: NVIDIA A40
CUDA version: 12.4
PyTorch version: 2.3.0
Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]


D as tensor, torch.compile off:
N=1024: 1.58 unit/s
N=2048: 1.43 unit/s
N=3072: 1.35 unit/s
N=4096: 1.32 unit/s

D as list, torch.compile off:
N=1024: 1.41 unit/s
N=2048: 1.27 unit/s
N=3072: 1.23 unit/s
N=4096: 1.20 unit/s
N=5120: 1.19 unit/s

D as tensor, torch.compile on (also takes a long time to compile for the first matrix shape, not shown here):
N=1024: 0.71 unit/s
N=2048: 1.51 unit/s
N=3072: 1.46 unit/s
N=4096: 1.42 unit/s
N=5120: 1.40 unit/s
N=6144: 1.37 unit/s

D as list, torch.compile on (also takes a long time to compile for each matrix shape, not shown here):
N=1024: 0.67 unit/s
N=2048: 0.60 unit/s
N=3072: 0.56 unit/s
(I stop here, the torch.compiled D list are taking too long)
```

**Comment by ad8e** [comment]
This is obsolete now that AI can write kernels. Want to close it?
